Weekly_Work/Week2/Code_Changes_Git.py

```diff
@@ -0,0 +1,159 @@
+            
+            if master_only:
+                count_per_repo[repo_url] = self._count_master_only(repo_url, empty)
+
+            else:
+                if not empty:
+                    df = df[df['files_action'] != 0]
+                if not merge:
+                    df = df[df['merge'] == False]
+                count_per_repo[repo_url] = df['hash'].nunique()
+            
+        return count_per_repo
+    
+    def compute_timeseries(self, period="month", plot_chart=False):
+        """
+        The metric value is computed for each fixed interval of time
+        from the "since" date to the "until" date arguments, specified 
+        during object initiation.
+        
+        The fixed time interval can be either a month or a week.
+        
+        :param period: A string which can be either "month" or "week"
+        :param plot_chart: Plots a barchart of the timeseries if True
+        """
+        dataframe_per_repo = dict()
+        for repo_url in self.repo_urls:
+            
+            df = self.df[self.df['repo'] == repo_url]
+            if period == "month":
+                timeseries_series =  df['created_date'] \
+                    .groupby([df['created_date'].dt.year.rename('year'),
+                              df['created_date'].dt.month.rename('month')]) \
+                    .agg('count')
+
+                all_periods = pd.DataFrame(pd.date_range(self.since, self.until, freq='M'), columns=["Dates"])
+                all_periods = pd.DataFrame(    \
+                    [all_periods['Dates'].dt.year.rename("year"),   \
+                     all_periods['Dates'].dt.month.rename("month")]).T
+
+            elif period == "week":
+                timeseries_series =  df['created_date'] \
+                    .groupby([df['created_date'].dt.year.rename('year'),
+                              df['created_date'].dt.week.rename('week')]) \
+                    .agg('count') 
+
+                all_periods = pd.DataFrame(pd.date_range(self.since, self.until, freq='W'), columns=["Dates"])
+                all_periods = pd.DataFrame(    \
+                    [all_periods['Dates'].dt.year.rename("year"),   \
+                     all_periods['Dates'].dt.week.rename("week")]).T
+
+
+            timeseries_df = pd.DataFrame(timeseries_series)
+            timeseries_df.reset_index(inplace=True)
+            timeseries_df.columns = ["year", period, "count"]
+            merged_df = all_periods.merge(timeseries_df, how='outer').fillna(0)
+
+            if plot_chart == True:
+                plt.style.use('seaborn')
+                merged_df.plot(y='count', use_index=True)
+                plt.fill_between(y1=merged_df['count'], y2=0, x=merged_df.index)
+                plt.title(repo_url)
+                plt.show()
+            
+            dataframe_per_repo[repo_url] = merged_df
+        return dataframe_per_repo
+    
+    def _count_master_only(self, repo_url, empty=False):
```

Comment:
Shouldn't this be just a filter?

Reply:
It is a filter. The snippet for finding only the master commits is quite long. I made it into a method which should n't be disclosed (hence the `_`). This method is called in `compute` and returns the count of master commits. In `compute`, there is a `master_only` parameter, which when True, directly returns `_count_master_only`:
```python
if master_only:
     count_per_repo[repo_url] = self._count_master_only(repo_url, empty)
else: ...
     ...
return count_per_repo
```

Will remove the per-repo count is the next patch.